### ISSUE TITLE
Deduplicate nodes in Expand's output

### ISSUE THREAD

**Issue by amykglen** [RTXteam/RTX#823]
Need to start deduplicating nodes in Expand's output, using @edeutsch's future KGNodeIndex method (based on the SRI team's node normalizer).

(Note: Currently, no deduplication is done for KG1/KG2 Expands. Some deduplication is done for BTE Expands, but the current method for choosing the [preferred identifier](https://github.com/RTXteam/RTX/blob/851c2b5b350e0350d76d2824f7c6683f6cea6a4d/code/ARAX/ARAXQuery/Expand/expand_utilities.py#L105) is quite hacky - not ideal.)

**Comment by amykglen** [comment]
Been starting to plan out how to implement this in `expand`, and am unsure about how to handle a couple situations...

For both, consider a query such as this one:
```
(n00: MONDO:0005180)--(n01: protein)
```
And assume the following:

- Synonyms are used for `n00` with `map_back` synonym handling
- The synonym group for `MONDO:0005180` looks like this: `[MONDO:0005180, DOID:14330, CUI:C0030567]`
- The 'preferred identifier' for that synonym group is `DOID:14330` (according to whatever future method we will have in KGNodeIndex that decides preferred identifiers)

Then my questions are as follows:

1. If `expand` runs this query and gets answers back from the KP, some of which use the MONDO node and some of which use the DOID node as `n00`, should the node corresponding to `n00` that `expand` ultimately returns be the `DOID:14330` one (since it is 'preferred'), even though the query graph came in using the MONDO curie specifically?  (my vote: yes)
2. If `expand` runs this query and gets answers back from the KP, all of which use `CUI:C0030567` as `n00` (so the MONDO and DOID nodes aren't returned from the KP at all), how should `expand` create the `n00` node it will add to the KG? Should it 'repurpose' the CUI node returned from the KP by just changing its curie to the 'preferred identifier'? Or should it create a sort of 'shell' of a node that has only a curie, name, and type (obtained from KGNodeIndex)?

(Note: I'm focusing on the 'input' node here just because it makes for a simpler example, but the answers to these questions will inform how to handle the 'output' nodes too...)

**Comment by dkoslicki** [comment]
Re:
>If expand runs this query and gets answers back from the KP, some of which use the MONDO node and some of which use the DOID node as n00, should the node corresponding to n00 that expand ultimately returns be the DOID:14330 one (since it is 'preferred'), even though the query graph came in using the MONDO curie specifically?

From what I've seen in the Architecture calls, yes, you should return the "preferred" curie even if the QG asked for something else. The hope is that the ARS will only be asking QG's with preferred curies anyways, so hopefully will be a non-issue for those accessing the system (ultimately) via the ARS.

>If expand runs this query and gets answers back from the KP, all of which use CUI:C0030567 as n00 (so the MONDO and DOID nodes aren't returned from the KP at all), how should expand create the n00 node it will add to the KG? Should it 'repurpose' the CUI node returned from the KP by just changing its curie to the 'preferred identifier'? Or should it create a sort of 'shell' of a node that has only a curie, name, and type (obtained from KGNodeIndex)?

AFAIK, the plan is to use the SRI developed [normalizer/synonymizer](https://nodenormalization-sri.renci.org/apidocs/), but it's still in flux as to when this will be applied (during ARS->ARA, ARA->KP, KP->ARA, ARA->ARS, etc.). So FWIW, my thought is: do what you want for now, and wait to hear "from on high" (i.e. the SRI group) about when and where to apply the normalizer and map back/merge/repurposing issue.

Note that the normalizer currently says the MONDO curie is currently preferred (see https://nodenormalization-sri.renci.org/get_normalized_nodes?curie=DOID%3A14330), but (as it's still a work in progress, doesn't know what to do with CUI:C0030567).

@edeutsch feel free to weigh in

**Comment by edeutsch** [comment]
Not much more to add, I think this is all right or TBD. Hopefully the new KGNodeIndex will have all this information at your fingertips. You'll be able to check what the SRI normalizer has to say and what our system recommends, what nodes we have, etc. So whatever is decided in the end would be an easy code change to conform to that. KG2NodeIndex is on day 2 at 24% of nodes organized. Still a ways to go.

**Comment by amykglen** [comment]
An update here: This has been implemented in the `expander` branch using a temporary way of determining the 'preferred curie' for a node (to be swapped out for the new KGNodeIndex synonymizer methods when they're ready).

I was figuring I will wait to get this merged into `master` until the new synonymizer is ready, though I suppose we don't have to since even without those methods, this still represents an improvement over the current `expand`. There will be a few things to work out with other modules before merging these changes, however (whenever that might be) - specifically, these pytests are unhappy with the changes:
```
FAILED test_ARAX_filter_kg.py::test_default_std_dev - assert 42 == 21
FAILED test_ARAX_overlay.py::test_FET_ex1 - AssertionError: assert 'ERROR' == 'OK'
FAILED test_ARAX_workflows.py::test_FET_example_1 - AssertionError: assert 'ERROR' == 'OK'
FAILED test_ARAX_workflows.py::test_FET_example_2 - AssertionError: assert 'ERROR' == 'OK'
FAILED test_ARAX_workflows.py::test_FET_example_4 - AssertionError: assert 'ERROR' == 'OK'
FAILED test_ARAX_workflows.py::test_clinical_overlay_example - AssertionError: assert 1 >= 2
```
I believe some are unhappy because they rely on `expand` returning a specific number of nodes/edges, and those numbers have changed thanks to deduplication. The FET tests seem more deeply unhappy though, which I suspect is related to the fact that the 'preferred curie' for a given node isn't necessarily present in the KP that was used... (though I'm not sure if that is the problem). @chunyuma

(as a side note: it might be wise for our pytests to, where possible, avoid having a dependency on the number of nodes/edges `expand` outputs... not only because of situations such as this one where `expand` is improved, but also because those numbers could change with each KG2 build that's rolled out, each improvement to the synonymizer that's rolled out, etc..)

**Comment by chunyuma** [comment]
@amykglen, sorry for the late reply. I just tested it and all the `FET` test within `test_ARAX_workflows.py` work for me. I'm not sure if something updated fixes this.

**Comment by amykglen** [comment]
no worries, @chunyuma - this isn't a rush at all because this is about changes in the `expander` branch for a new enhancement (not in `master`) - which may be why you didn't see failing tests. did you run them in the `expander` branch?

**Comment by chunyuma** [comment]
@amykglen, oh sorry. I just ran it in `master`. I will try to run it in the `expander` branch and will give you the feedback.

**Comment by chunyuma** [comment]
@amykglen, you're right. They failed because of the 'preferred curie' problem. Some `preferred curie` nodes might only exist in KG2 but not KG1. To fix it, I think I need a way to map the `preferred curie` back to the one that exist in KG1. But we might not know which one is the one that originally exist in KG1, right?

**Comment by amykglen** [comment]
ok, got it. yeah - one solution is to do what `expand` does and use the KGNodeIndex method `get_equivalent_curies()` to get a list of the 'synonyms' for that curie, and then use those in your cypher query in place of the 'preferred curie'.

but instead of that, it might make more sense to just use `expand` to answer the query, because `expand` already automatically knows to use those synonyms and how to deduplicate the answers. is that an option to use `expand` rather than your own cypher query?

**Comment by chunyuma** [comment]
@amykglen, yes, for KG1, there is an option to use `expand` or `cypher query`. But currently `FET` takes cypher query as default for KG1 because it is much faster.  For example, `test_FET_example_2` within `test_ARAX_workflow.py` costs a few seconds when using cypher query but needs around 10 minutes when using `expand`. 

If this is the case, then KG2 should not have this problem because it always uses `expand`.

So perhaps using the KGNodeIndex method `get_equivalent_curies()` to get a list of the 'synonyms' for that curie might be faster for KG1. But do I need to check the synonyms one by one from the returned list to see which one is in KG1?

**Comment by amykglen** [comment]
@chunyuma - interesting. When you do the method that uses Expand, do you also call Resultify?

**Comment by chunyuma** [comment]
@amykglen, yes, I always used `Resultify` in DSL within `FET`. I think I should not use `Resultify` in `FET` because it might consume time, right? Good reminder. Thanks!

**Comment by amykglen** [comment]
Right. Nice! Yeah, I’d try getting rid of the Resultify call, and once you do, I think the time when using the Expand method should be about the same as with your cypher query. 

It would definitely be ideal if you could use Expand I think, because synonymizing/deduplicating nodes isn’t a trivial thing to do, and Expand takes care of all that for you.

Let me know if time is still an issue after you remove Resultify!

**Comment by chunyuma** [comment]
This DSL might be a good example for the future test for the new KGNodeIndex.
```
add_qnode(curie=UniProtKB:P06724, id=n00, type=protein)
add_qnode(type=chemical_substance, is_set=true, id=n01)
add_qedge(source_id=n00, target_id=n01,id=e00)
expand(edge_id=e00, kp=ARAX/KG1)
```

**Comment by amykglen** [comment]
thanks, @chunyuma!

for the record: Chunyu and I worked out the problems FET was having with deduplication/preferred curies, so now all FET tests are passing in the `expander` branch except for `test_FET_example_2` in `test_ARAX_workflows.py`, which is failing due to the KGNodeIndex bug in #825 (specifically for protein `UniProtKB:P06724`, which Chunyu posted above). once the new KGNodeIndex synonymizer is ready, I'll plug that into `expand` and we can verify that this FET test is passing as well.

**Comment by amykglen** [comment]
confirmed `test_FET_example_2` is passing after plugging in the new `NodeSynonymizer` (as well as all other FET tests). :)

**Comment by amykglen** [comment]
played around on `/beta` to test this - all looks good
